yaml/shared/src/main/scala/org/virtuslab/yaml/internal/load/reader/Tokenizer.scala

```diff
@@ -4,35 +4,38 @@ import org.virtuslab.yaml.internal.load.reader.token.{BlockChompingIndicator, Sc
 import org.virtuslab.yaml.internal.load.reader.token.Token
 import org.virtuslab.yaml.internal.load.reader.token.TokenKind.*
 import org.virtuslab.yaml.internal.load.reader.token.BlockChompingIndicator.*
+import org.virtuslab.yaml.ParseError
 
 import scala.util.Try
 import scala.annotation.tailrec
-import org.virtuslab.yaml.Position
+import org.virtuslab.yaml.{Position, YamlError}
 
 trait Tokenizer:
-  def peekToken(): Token
+  def peekToken(): Either[YamlError, Token]
   def popToken(): Token
 
 private[yaml] class Scanner(str: String) extends Tokenizer {
 
   private val ctx = ReaderCtx(str)
   private val in  = ctx.reader
 
-  override def peekToken(): Token = ctx.tokens.headOption match
-    case Some(token) => token
+  override def peekToken(): Either[YamlError, Token] = ctx.tokens.headOption match
+    case Some(token) => Right(token)
     case None        => getToken()
 
   override def popToken(): Token = ctx.tokens.removeHead()
 
-  private def getToken(): Token =
-    ctx.tokens.appendAll(getNextTokens())
-    ctx.tokens.head
+  private def getToken(): Either[YamlError, Token] =
+    getNextTokens().map { tokens =>
+      ctx.tokens.appendAll(tokens)
+      ctx.tokens.head
+    }
 
-  private def getNextTokens(): List[Token] =
+  private def getNextTokens(): Either[YamlError, List[Token]] =
     skipUntilNextToken()
     val closedTokens = ctx.checkIndents(in.column)
     val peeked       = in.peek()
-    val tokens = peeked match
+    val tokens: Either[YamlError, List[Token]] = peeked match
```

Comment:
Crazy idea, wdyt about throwing an exception without stack trace and cathing it in `getToken()`? It would reduce the amount of `Right(...)` needed to be written and probably save a few allocations.

Reply:
or if we want to avoid exceptions, I think that union type `YamlError | List[Token]` is also a promising option.